upload/views/combined_upload.py

```diff
@@ -0,0 +1,171 @@
+                is_shelter_request=self.is_shelter_request(),
+                position="start",
+            ),
+        )
+
+        # Create commit
+        create_commit_data = dict(
+            commitid=request.data.get("commit_sha"),
+            parent_commit_id=request.data.get("parent_sha"),
+            pullid=request.data.get("pull_request_number"),
+            branch=request.data.get("branch"),
+        )
+        commit_serializer = CommitSerializer(data=create_commit_data)
+        if not commit_serializer.is_valid():
+            return Response(
+                commit_serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+        log.info(f"Creating commit for {commit_serializer.validated_data}")
+        repository = self.get_repo()
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="create_commit",
+            ),
+        )
+        commit = self.create_commit(commit_serializer, repository)
+
+        # Create report
+        commit_report_data = dict(
+            code=request.data.get("code"),
+        )
+        commit_report_serializer = CommitReportSerializer(data=commit_report_data)
+        if not commit_report_serializer.is_valid():
+            return Response(
+                commit_report_serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="create_report",
+            ),
+        )
+        report = self.create_report(commit_report_serializer, repository, commit)
+
+        # Do upload
+        upload_data = dict(
+            ci_url=request.data.get("build_url"),
+            env=request.data.get("env_vars"),
+            flags=request.data.get("flags"),
+            ci_service=request.data.get("ci_service"),
+            job_code=request.data.get("job_code"),
+            name=request.data.get("name"),
+        )
+
+        upload_serializer = UploadSerializer(data=upload_data)
+        if not upload_serializer.is_valid():
+            return Response(
+                upload_serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        inc_counter(
+            API_UPLOAD_COUNTER,
+            labels=generate_upload_prometheus_metrics_labels(
+                action="coverage",
+                endpoint="combined_upload",
+                request=self.request,
+                is_shelter_request=self.is_shelter_request(),
+                position="create_upload",
```

Comment:
these metrics take up half the of the function, and they only differ in `position`.

Can you extract these into a separate function `emit_metric(self, position: str)`, so that the actual function logic is more visible?

Reply:
Good point! This should also be fixed, there is now an `emit_metrics` function to handle this.